python/lib/modeldata/dmod/modeldata/subset/hydrofabric_subset.py

```diff
@@ -127,17 +133,12 @@ def factory_create_from_base_and_hydrofabric(cls, subset_def: SubsetDefinition,
         return cls(catchment_ids=subset_def.catchment_ids, nexus_ids=subset_def.nexus_ids, hydrofabric=hydrofabric,
                    *args, **kwargs)
 
-    __slots__ = ["_catchments", "_nexuses"]
-
-    def __init__(self, catchment_ids: Collection[str], nexus_ids: Collection[str], hydrofabric: Hydrofabric, *args,
-                 **kwargs):
-        self._catchments: Set[Catchment] = set()
-        self._nexuses: Set[Nexus] = set()
-        super().__init__(catchment_ids, nexus_ids, hydrofabric)
+    def __init__(self, catchment_ids: Collection[str], nexus_ids: Collection[str], hydrofabric: Hydrofabric, **data):
+        super().__init__(catchment_ids=catchment_ids, nexus_ids=nexus_ids, hydrofabric=hydrofabric, **data)
         # Since super __init__ validates, and validate function make sure ids are recognized, these won't ever be None
-        for cid in catchment_ids:
+        for cid in self.catchment_ids:
             self._catchments.add(hydrofabric.get_catchment_by_id(cid))
-        for nid in nexus_ids:
+        for nid in self.nexus_ids:
             self._nexuses.add(hydrofabric.get_nexus_by_id(nid))
```

Comment:
Yeah, I could go either way on the loops-versus-comprehension choice here. If I have to pick, I say leave it because it's already there.

> Also, it looks like `_catchments` and `_nexuses` aren't meant to be modified once created. Would it make sense to store them as tuples instead of creating new tuples in their property getters?

Hmm, we'd want to maintain something that protects from duplicates.  Also, I'm noticing now that we've modified `validate_hydrofabric` in a way that no longer benefits from the backing data structures being `set` objects.  We might want to think about that for a bit, and whether it's worth reintroducing, before changing everything to `tuple`.  There is definitely a trade-off for having to re-create a `tuple` each property invocation; I'm not sure right away which is the better optimization.

Reply:
I'm 110% on board with Austin's point about array allocations.

Reply:
> I'm noticing now that we've modified validate_hydrofabric in a way that no longer benefits from the backing data structures being set objects

I realize that I thought about this incorrectly before.  We aren't actually checking if an id is in the data structures; we are just iterating through the data structures.  So the tuple would actually be faster too.  

With that said, I think we do need to abandon the loops and do something like this:

``` Python
self._catchments = tuple(hydrofabric.get_catchment_by_id(cid) for cid in self.catchment_ids)
self._nexuses = tuple(hydrofabric.get_nexus_by_id(nid) for nid in self.nexus_ids)
```

We could also consider changing the attributes from private to regular fields, ditching the property functions, and doing the above initialization inside validators.  Though that may be excessive.

Thoughts or alternatives @aaraney and @christophertubbs?

Reply:
That's fine with me, @robertbartel. Just so we are on the same page, the `_catchments` and `_nexuses` attributes dont need to guarantee uniqueness?